repl/src/main/scala/com/cloudera/livy/repl/SparkSqlInterpreter.scala

```diff
@@ -0,0 +1,301 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.cloudera.livy.repl
+
+import java.io._
+
+import scala.tools.nsc.Settings
+import scala.tools.nsc.interpreter.{JPrintWriter, Results}
+import scala.util.{Failure, Success, Try}
+
+import org.apache.spark.{SparkConf, SparkContext}
+import org.apache.spark.rdd.RDD
+import org.apache.spark.repl.SparkIMain
+import org.apache.spark.sql.hive.HiveContext
+import org.json4s.{DefaultFormats, Extraction}
+import org.json4s.JsonAST._
+import org.json4s.JsonDSL._
+
+object SparkSqlInterpreter {
+  private val MAGIC_REGEX = "^%(\\w+)\\W*(.*)".r
+}
+
+/**
+ * This represents a Spark Sql interpreter. It is not thread safe.
+ */
+class SparkSqlInterpreter(conf: SparkConf) extends Interpreter {
+  import SparkSqlInterpreter._
+
+  private implicit def formats = DefaultFormats
+
+  private val outputStream = new ByteArrayOutputStream()
+  private var sparkIMain: SparkIMain = _
+  private var sparkContext: SparkContext = _
+  private var hiveContext: HiveContext = _
+
+  def kind: String = "sparksql"
+
+  override def start(): SparkContext = {
+    require(sparkIMain == null && sparkContext == null && hiveContext == null)
+
+    val settings = new Settings()
+    settings.embeddedDefaults(Thread.currentThread().getContextClassLoader())
+    settings.usejavacp.value = true
+
+    sparkIMain = new SparkIMain(settings, new JPrintWriter(outputStream, true))
+    sparkIMain.initializeSynchronous()
+
+    // Spark 1.6 does not have "classServerUri"; instead, the local directory where class files
+    // are stored needs to be registered in SparkConf. See comment in
+    // SparkILoop::createSparkContext().
+    Try(sparkIMain.getClass().getMethod("classServerUri")) match {
+      case Success(method) =>
+        method.setAccessible(true)
+        conf.set("spark.repl.class.uri", method.invoke(sparkIMain).asInstanceOf[String])
+
+      case Failure(_) =>
+        val outputDir = sparkIMain.getClass().getMethod("getClassOutputDirectory")
+        outputDir.setAccessible(true)
+        conf.set("spark.repl.class.outputDir",
+          outputDir.invoke(sparkIMain).asInstanceOf[File].getAbsolutePath())
+    }
+
+    restoreContextClassLoader {
+      // Call sparkIMain.setContextClassLoader() to make sure SparkContext and repl are using the
+      // same ClassLoader. Otherwise if someone defined a new class in interactive shell,
+      // SparkContext cannot see them and will result in job stage failure.
+      val setContextClassLoaderMethod = sparkIMain.getClass().getMethod("setContextClassLoader")
+      setContextClassLoaderMethod.setAccessible(true)
+      setContextClassLoaderMethod.invoke(sparkIMain)
+
+      sparkContext = SparkContext.getOrCreate(conf)
+      hiveContext = new HiveContext(sparkContext)
+
```

Comment:
I think sometimes just SQLContext is sufficient, especailly when there's no hive-site.xml on classpath. Link with #132 which is related .